api/v1/deployment/deployer.go

```diff
@@ -3,37 +3,43 @@ package deployment
 import (
 	"encoding/json"
 
+	"github.com/jinzhu/gorm"
 	"github.com/juju/errors"
+	"github.com/ovh/lhasa/api/db"
 	"github.com/ovh/lhasa/api/v1"
 )
 
 // Deployer deploys an application version to the given environment and removes old deployments
 type Deployer func(application v1.Release, environment v1.Environment, deployment *v1.Deployment) error
 
 // ApplicationDeployer deploys an application version to the given environment and removes old deployments
-func ApplicationDeployer(depRepo *Repository) Deployer {
+func ApplicationDeployer(tm db.TransactionManager, depFactory RepositoryFactory) Deployer {
 	return func(app v1.Release, env v1.Environment, dep *v1.Deployment) error {
-		if err := depRepo.UndeployByApplicationEnv(app.Domain, app.Name, env.ID); err != nil {
-			return err
-		}
-		props := map[string]interface{}{}
-		if len(dep.Properties.RawMessage) > 0 {
-			if err := json.Unmarshal(dep.Properties.RawMessage, &props); err != nil {
-				return errors.BadRequestf("properties field should be a valid json object: %s", err.Error())
+		return tm.Transaction(func(db *gorm.DB) error {
+			depRepo := depFactory(db)
+
+			if err := depRepo.UndeployByApplicationEnv(app.Domain, app.Name, env.ID); err != nil {
+				return err
+			}
+			props := map[string]interface{}{}
+			if len(dep.Properties.RawMessage) > 0 {
+				if err := json.Unmarshal(dep.Properties.RawMessage, &props); err != nil {
```

Comment:
You can squeeze the two conditions in a single one.

Reply:
I don't see how ?

Reply:
My bad